core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala

```diff
@@ -109,13 +112,62 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
       _ = logger.info(
         s"Deleting workspace [workspaceId=${workspace.workspaceId}, name=${workspaceName.name}, billingProject=${workspace.namespace}, user=${ctx.userInfo.userSubjectId.value}]"
       )
-      result: Option[String] <- workspace.workspaceType match {
+      result: WorkspaceDeletionResult <- workspace.workspaceType match {
         case RawlsWorkspace => workspaceService.deleteWorkspace(workspaceName)
         case McWorkspace    => deleteMultiCloudWorkspace(workspace)
       }
+    } yield result.gcpContext.flatMap(_.bucketName)
+
+  def deleteMultiCloudOrRawlsWorkspaceV2(workspaceName: WorkspaceName,
+                                         workspaceService: WorkspaceService
+  ): Future[WorkspaceDeletionResult] =
+    for {
+      workspace <- getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.delete)
+
+      _ = logger.info(
+        s"V2 Starting deletion of workspace [workspaceId=${workspace.workspaceId}, name=${workspaceName.name}, billingProject=${workspace.namespace}, user=${ctx.userInfo.userSubjectId.value}]"
+      )
+
+      _ = workspace.state match {
+        case WorkspaceState.Ready => true
+        case WorkspaceState.Deleting =>
+          throw new RawlsExceptionWithErrorReport(
+            ErrorReport(StatusCodes.Conflict, "Workspace is already being deleted.")
+          )
+        case _ =>
```

Comment:
Do we need to allow workspaces in `CreateFailed`, `DeleteFailed`, and `UpdateFailed` to be deleted so users can clean up their failed workspaces?

Reply:
That's a good point.

I'm trying to think through the implications of that. We might need to make some steps more resilient to things like incomplete records or resources, for that to work.
On the other hand, if deleting doesn't work, it just goes from one failed state to another? 

Maybe we should allow this after the 'Nuke it from orbit' plan is in place?

Reply:
I think we should allow deletion for any state besides Deleting.

Reply:
I will update. We don't want to admit transitions from CREATING and UPDATING (per the [design spec](https://broadworkbench.atlassian.net/wiki/spaces/WOR/pages/2819653639/Workspace+Async+Deletion)) so I will fail those out.